WooCommerce: Add fields for simple product type

justinshreve opened the pull request:
This adds the other product configuration fields that should be present when a product is configured as a 'simple' product (non-variable).

This includes price input, weight, and stock settings.

There are two `TODO`s in this file: The currency, currency position, dimensions unit, and weight unit all need to be passed in to the product form still (they are currently hardcoded). I have opened #14185 for the product form as a whole, so we make sure to come back and handle these.

cc @kellychoffman @jameskoster as this contains UI elements.
Closes #13664, #13665, #13664.

To test:
* Go to `http://calypso.localhost:3000/store/products/:site/add`.
* Toggle to variations and make sure those cards disappear and the variations UI is displayed instead. 'Simple' fields are dropped from state (like how we drop variation attributes when switching to simple). Toggle the variations switch again and make sure you can see the simple fields again.
* Configure the following various product settings and make sure their values update: Price, dimensions, weight, stock (checkbox, quantity), and backorders.

<img width="713" alt="screen shot 2017-05-17 at 10 20 41 am" src="https://cloud.githubusercontent.com/assets/689165/26167487/c012cf9e-3aeb-11e7-993b-680fe55eef9a.png">

a8ci18n commented on client/extensions/woocommerce/app/products/product-form-simple-card.js:
```diff
@@ -0,0 +1,150 @@
+					/>
+				</FormFieldSet>
+				<FormFieldSet>
+					<FormLabel>{ translate( 'Weight' ) }</FormLabel>
```
Hi! I've found a possible matching string that has already been translated **24** times:
`translate( 'Weight', { context: 'Dns Record'} )` *ES Score: 19*
See 1 additional suggestion in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/add/woocommerce-simple-product-fields)

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

kellychoffman commented:
I am getting a bunch of errors when I `make run` and it doesn't render the page. Let me know if you want me to paste the errors.

justinshreve commented:
> I am getting a bunch of errors when I make run and it doesn't render the page. Let me know if you want me to paste the errors.

Doh. A change I thought I made didn't get committed so it was trying to load the wrong file. Not sure how I missed that. I have amended my commit and it should build now. @kellychoffman Can you pull down the changes and try again?

jameskoster commented:
Made a couple of minor tweaks, looking really good!

kellychoffman commented:
Can you add a 'Product price' label above the price input?

kellychoffman commented:
@jameskoster - Is the mobile view broken for you? The sidebar doesn't display on this page for me, so it could be a byproduct of that.

kellychoffman commented:
The interaction of the manage stock / backorders feels odd. One thing we should improve to make this feel smoother: 

I think the quantity field should always be visible. If you enter a quantity, the checkbox is checked.

kellychoffman commented:
Made some small changes: price placeholder and more space in between dimensions and weight. I think this is good to go from a design standpoint unless the mobile is really broken, but I have a feeling thats just on my end.

justinshreve commented:
Setting back to `In Progress` until I have a chance to change the stock behavior.

jameskoster commented:
> I think the quantity field should always be visible. If you enter a quantity, the checkbox is checked.

I think it might feel stranger this way, personally :p Keen to see it live. But as Justin said on Slack, it makes the checkbox questionable. To make it clear the field is optional couldn't we put something in the placeholder? "Leave blank if you don't manage stock". Then just lose the checkbox. Reveal Backorder option if the Manage stock input isn't empty.

Looks like I missed a couple of issues with the dimensions input on really small screens. Fixing now.

a8ci18n commented on client/extensions/woocommerce/app/products/product-form-simple-card.js:
```diff
@@ -0,0 +1,151 @@
+					/>
+				</FormFieldSet>
+				<FormFieldSet className="products__product-weight-input">
+					<FormLabel>{ translate( 'Weight' ) }</FormLabel>
```
Hi! I've found a possible matching string that has already been translated **24** times:
`translate( 'Weight', { context: 'Dns Record'} )` *ES Score: 19*
See 1 additional suggestion in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/add/woocommerce-simple-product-fields)

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

jameskoster commented:
OK there were a few more things than I spotted first time around! Mostly done, couple of small things outstanding;

![stop-wrap](https://cldup.com/p_fgYNrKm9-3000x3000.png)

We need to stop these inputs wrapping in the variations table. As we have horizontal scrolling it's not necessary. I was able to do this for the currency input [using the `noWrap` flag](https://github.com/Automattic/wp-calypso/pull/14186/commits/73f60dbafe9761a49439e673b7d4e84d2ed52ec3#diff-475180105f8fc641fc846aedbf5c70d4R73).

I can't get it to work properly for the dimensions / weight though. Obviously we'll need to add some styling for the dimensions as this is a new component. But I couldn't get the class added to the weight input either. I was trying to add the `noWrap` flag [here](https://github.com/Automattic/wp-calypso/blob/73f60dbafe9761a49439e673b7d4e84d2ed52ec3/client/extensions/woocommerce/app/products/product-form-variations-row.js#L90) but no dice. So yeah, we need the `no-wrap` class added to the weight input and the dimensions inputs.

justinshreve commented:
> I think it might feel stranger this way, personally :p Keen to see it live. But as Justin said on Slack, it makes the checkbox questionable. To make it clear the field is optional couldn't we put something in the placeholder? "Leave blank if you don't manage stock". Then just lose the checkbox. Reveal Backorder option if the Manage stock input isn't empty.

How about we keep stock as-is for this PR, and then I can throw up a try branch so we can look at how it looks.

justinshreve commented:
@jameskoster I handled the noWrap stuff.

I'm going to set this back to "Needs Review" since I will plan on playing with the stock input in a separate PR.

jameskoster commented:
> I handled the noWrap stuff.

Thanks!

> How about we keep stock as-is for this PR, and then I can throw up a try branch in a bit so we can look at how it looks.

Works for me.

allendav commented:
If you enter a negative number in the stock input, you get a warning:

<img width="938" alt="screen shot 2017-05-19 at 11 26 30 am" src="https://cloud.githubusercontent.com/assets/1595739/26261464/1ffa49c8-3c86-11e7-8f54-7a577dc9cfb3.png">

happens for weight field too

allendav commented on client/extensions/woocommerce/app/products/product-form-variations-card.js:
```diff
@@ -22,7 +22,21 @@ const ProductFormVariationsCard = ( {
 	const handleToggle = () => {
 		if ( 'variable' !== product.type ) {
-			editProduct( product, { type: 'variable', dimensions: {} } );
+			// Reset any 'simple' product settings that have been filled out,
```
Eeek! Are you sure you want to do this?

allendav replied:
Maybe save the values (if any) in component state and put them back if they toggle it off again?

justinshreve replied:
We had decided at some point to clear out variation types when the type gets switched back to simple, so if we do that, we should do the same in reverse.

However - I agree with you that it's a little too easy to lose state by toggling that. I'm not sure what the answer is though, because we don't want to pass that extra data to the API.

justinshreve replied:
Actually I guess I could store that data in local state and reapply it to the Redux state when they toggle. 

Maybe it's a good idea to remove these lines for now and do that in a separte PR.

justinshreve replied:
*shakes fist*. GitHub's UI didn't show me your second comment. Agreed. I think this PR is large enough so I can remove the toggle logic here, and be smarter about this in a future PR.

justinshreve commented:
Looks like price, weight, etc are actually returned strings via the API anyway, so I changed the PropType. I also prevented negative values from being saved to state in the first place since they are invalid. The stock input will be changing a bit more in an upcoming PR too.

Added a TODO for the toggle and backed out those changes.

@allendav Good to go?

kellychoffman commented:
> The stock input will be changing a bit more in an upcoming PR too.

Just to clarify: does this mean you are making the 'always there' quantity input change in another PR?

justinshreve commented:
> Just to clarify: does this mean you are making the 'always there' quantity input change in another PR?

Yeah.

kellychoffman commented:
Awesome, :shipit:

allendav commented:
:shipit:

coderkevin commented on client/extensions/woocommerce/app/products/product-form-simple-card.js:
```diff
@@ -0,0 +1,155 @@
+					{ product.manage_stock && (
+						<FormTextInput
+							name="stock_quantity"
+							value={ product.stock_quantity || '' }
```
Should this default to zero?

justinshreve replied:
@coderkevin Yeah it should have. However, I already have a `/try` branch open that changes the behavior of the stock inputs a bit based on design feedback.

coderkevin commented on client/extensions/woocommerce/app/products/product-form-simple-card.js:
```diff
@@ -0,0 +1,155 @@
+			{ renderDeliveryDetails() }
+			{ renderStock() }
+		</div>
+	);
```
I like how simple the final render looks here.
Just keep in mind if these things are broken out, you'll have to pass them parameters then.
